cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java

```diff
@@ -101,25 +126,108 @@ private void refreshCredentials() throws IOException {
             .build();
 
     StsTokenExchangeResponse response = handler.exchangeToken();
-    this.accessBoundarySessionKey = response.getAccessBoundarySessionKey();
-    this.intermediateAccessToken = response.getAccessToken();
 
-    // The STS endpoint will only return the expiration time for the intermediary token
+    synchronized (refreshLock) {
+      this.accessBoundarySessionKey = response.getAccessBoundarySessionKey();
+      this.intermediateAccessToken = getTokenFromResponse(response, sourceAccessToken);
+    }
+  }
+
+  private static AccessToken getTokenFromResponse(
+      StsTokenExchangeResponse response, AccessToken sourceAccessToken) {
+    AccessToken intermediateToken = response.getAccessToken();
+
+    // The STS endpoint will only return the expiration time for the intermediate token
     // if the original access token represents a service account.
-    // The intermediary token's expiration time will always match the source credential expiration.
+    // The intermediate token's expiration time will always match the source credential
+    // expiration.
     // When no expires_in is returned, we can copy the source credential's expiration time.
-    if (response.getAccessToken().getExpirationTime() == null) {
-      if (sourceAccessToken.getExpirationTime() != null) {
-        this.intermediateAccessToken =
-            new AccessToken(
-                response.getAccessToken().getTokenValue(), sourceAccessToken.getExpirationTime());
+    if (intermediateToken.getExpirationTime() == null
+        && sourceAccessToken.getExpirationTime() != null) {
+      return new AccessToken(
+          intermediateToken.getTokenValue(), sourceAccessToken.getExpirationTime());
+    }
+    return intermediateToken; // Return original if no modification needed
+  }
+
+  private void startAsynchronousRefresh() {
+    // Obtain the lock before checking or modifying currentRefreshFuture to prevent race conditions.
+    synchronized (refreshLock) {
```

Comment:
qq, why do we need to obtain the lock here. I think for this it's probably fine for async, but I'm not putting together why it's needed.

Reply:
The lock here prevents multiple concurrent asynchronous refreshes from starting. Without the lock, the following scenario could occur:
1. Thread A checks currentRefreshFuture and finds it null or done.
2. Before Thread A can assign a new future to currentRefreshFuture, Thread B also checks and finds it null or done.
3. Both Thread A and Thread B start a refresh, wasting resources and potentially leading to inconsistencies if they complete out of order and overwrite each other's results.

The lock ensures that only one asynchronous refresh can be in progress at any given time.

Reply:
Ok, I think I understand. I think there are a few things regarding this.

1. When either an async or sync call is made, we really only want one execution/ call to go through at a time. Even if `refreshCredentialsIfRequired()` is called repeatedly, then one call really should be made. For that, I believe we would want functionality similar to [this](https://github.com/googleapis/google-auth-library-java/blob/9a7c2e08a6fd891d802e3fb7729d163841bae524/oauth2_http/java/com/google/auth/oauth2/OAuth2Credentials.java#L259). It will find the existing execution or create a new one as a transaction. If multiple calls are made at the same time, each of the calls will find the same refresh action while it's still active.
2. For the sync flow, we should kick off the refresh action [here](https://github.com/googleapis/google-auth-library-java/blob/eb4b793cca84bd64c6de00db0303e9e10c566aae/cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java#L222) via the createRefreshTask linked above and block until the future is resolved via `blockingRefresh()`.
3. For the async flow, it should just create a refresh action via createRefreshTask